sdk/dotnet/Pulumi/Core/InputMap.cs

```diff
@@ -62,6 +62,31 @@ public void Add(string key, Input<V> value)
             set => Add(key, value);
         }
 
+        /// <summary>
+        /// Merge two instances of <see cref="InputMap{V}"/>. Returns a new <see cref="InputMap{V}"/>
+        /// without modifying any of the arguments. 
+        /// <para/>If both maps contain the same key, the value from the second map takes over.
+        /// </summary>
+        /// <param name="first">The first <see cref="InputMap{V}"/>. Has lower priority in case of
+        /// key clash.</param>
+        /// <param name="second">The second <see cref="InputMap{V}"/>. Has higher priority in case of
+        /// key clash.</param>
+        /// <returns>A new instance of <see cref="InputMap{V}"/> that contains the items from
+        /// both input maps.</returns>
+        public static InputMap<V> Merge(InputMap<V> first, InputMap<V> second)
+        {
+            var output = Output.Tuple(first._outputValue, second._outputValue)
+                               .Apply(dicts =>
+                               {
+                                   var result = new Dictionary<string, V>(dicts.Item1);
+                                   // Overwrite keys if duplicates are found
+                                   foreach (var (k, v) in dicts.Item2)
+                                       result[k] = v;
+                                   return result;
```

Comment:
i think you can literally just do:  `dicts => dicts.Item1.AddRange(dicts.Item2)` :)

Reply:
This will throw an exception if the keys are overlapping. I think throwing an exception from `Apply` isn't great, so I made them merge over (that's also how we use it in TS).

Reply:
Ugh.  You're definitely right.  How annoying!  :(

lgtm.